generative_models/DCGAN_Tensorflow-master/main.py

```diff
@@ -0,0 +1,131 @@
+noise_dim = 100
+
+# Saving Checkpoints
+checkpoint_dir = './training_checkpoints'
+checkpoint_prefix = os.path.join(checkpoint_dir, "ckpt")
+checkpoint = tf.train.Checkpoint(generator_optimizer=generator_optimizer,
+                                 discriminator_optimizer=discriminator_optimizer,
+                                 generator=generator,
+                                 discriminator=discriminator)
+
+# Defining Training Loop
+@tf.function
+def train_step(images):
+    noise = tf.random.normal([BATCH_SIZE, noise_dim])
+
+    with tf.GradientTape() as gen_tape, tf.GradientTape() as disc_tape:
+      generated_images = generator(noise, training=True)
+
+      real_output = discriminator(images, training=True)
+      fake_output = discriminator(generated_images, training=True)
+
+      gen_loss = generator_loss(fake_output)
+      disc_loss = discriminator_loss(real_output, fake_output)
+
+      gradients_of_generator = gen_tape.gradient(gen_loss, generator.trainable_variables)
+      gradients_of_discriminator = disc_tape.gradient(disc_loss, discriminator.trainable_variables)
+
+      generator_optimizer.apply_gradients(zip(gradients_of_generator, generator.trainable_variables))
+      discriminator_optimizer.apply_gradients(zip(gradients_of_discriminator, discriminator.trainable_variables))
+    return gen_loss,disc_loss
+
+def train(dataset, epochs):
+  for epoch in range(epochs):
+    gen_loss_list = []
+    disc_loss_list = []
+
+    for image_batch in dataset:
+      t=train_step(image_batch)
+      gen_loss_list.append(t[0])
+      disc_loss_list.append(t[1])
+
+    gen_loss = sum(gen_loss_list) / len(gen_loss_list)
+    disc_loss = sum(disc_loss_list) / len(disc_loss_list)
+
+    
+    print (f'Epoch {epoch+1}, gen loss={gen_loss},disc loss={disc_loss}')
+    
+    generate_and_save_images(generator,
+                             epoch + 1,
+                             seed)
+
+    # Save the model every 15 epochs
+    if (epoch + 1) % 15 == 0:
+      checkpoint.save(file_prefix = checkpoint_prefix)
+ 
+  return gen_loss_list,disc_loss_list            
+
+# Generate and save images
+def generate_and_save_images(model, epoch, test_input):
+  # Notice `training` is set to False.
+  # This is so all layers run in inference mode (batchnorm).
+  predictions = model(test_input, training=False)
+
+  fig = plt.figure(figsize=(4,4))
+
+  for i in range(predictions.shape[0]):
+      plt.subplot(4, 4, i+1)
+      plt.imshow(predictions[i, :, :, 0] * 127.5 + 127.5, cmap='gray')
+      plt.axis('off')
+
+  plt.savefig('image_at_epoch_{:04d}.png'.format(epoch))
+  plt.show()
+
+# Training our model
+plo=train(train_dataset, EPOCHS)
+
+# Plotting generator losses and discriminator losses
+plt.figure(figsize=(10,5))
+plt.title("Generator and Discriminator Loss During Training")
+plt.plot(plo[0],label="G")
+plt.plot(plo[1],label="D")
+plt.xlabel("iterations")
+plt.ylabel("Loss")
+plt.legend()
+plt.savefig('loss_plot.png')
+plt.show()

```

Comment:
move this function to utils.py

Reply:
which files do i need to move in assets folder

Reply:
all display files that used in readme, should be in assets (like result, plot, etc.)